selectel/dbaas_postgresql_utils.go

```diff
@@ -0,0 +1,85 @@
+package selectel
+
+import (
+	"context"
+	"errors"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/selectel/dbaas-go"
+)
+
+func resourceDBaaSPostgreSQLDatastoreV1PoolerFromSet(poolerSet *schema.Set) (*dbaas.Pooler, error) {
+	if poolerSet.Len() == 0 {
+		return nil, nil
+	}
+	var resourceModeRaw, resourceSizeRaw interface{}
+	var ok bool
+
+	resourcePoolerMap := poolerSet.List()[0].(map[string]interface{})
+	if resourceModeRaw, ok = resourcePoolerMap["mode"]; !ok {
+		return &dbaas.Pooler{}, errors.New("pooler.mode value isn't provided")
+	}
+	if resourceSizeRaw, ok = resourcePoolerMap["size"]; !ok {
+		return &dbaas.Pooler{}, errors.New("pooler.size value isn't provided")
+	}
+
+	resourceMode := resourceModeRaw.(string)
+	resourceSize := resourceSizeRaw.(int)
```

Comment:
NIT: `resourceSizeRaw` is `interface{}` type that is parsed to `resourcePoolerMap` from first element of a list generated from `poolerSet`.
Looks like there are a lot of places in this transformations that can lead to errors and maybe it is better to handle them and throw more descriptive errors.
e.g.: what if `poolerSet` will be empty?

Also is there a way to get rid of set struct for `poolerSet` and use map (or slice of maps at least) or something else? Because set looks redundant as far we are interested only in some "first" value from it that is used as a map.

Reply:
1. I'm not sure that poolerSet can become empty, because we have `Required: true` option for both pooler fields, so terraform is not gonna allow to create an empty set. The same logic can be applied for other fields that use sets

2. We can use map here, but I see a couple of problems here. Right now we use set for pooler settings for `selectel_dbaas_datastore_v1` and we can't change it because it's gonna break current code that uses sets. We can use maps for new datastore type specific resourses, though. Also it might become inconvenient for users if they need to use different structs for old and new resources

3. As I understand there is no way of adding validation for TypeMap in terraform. So, set is preferable here because it's easier to see available options and their restrictions for developers